lib/tree.rb

```diff
@@ -2,65 +2,339 @@ class TreeNode
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(currNode, list)
+
+    if currNode.nil?
+      # base case
+      return list
+
+    else
+      # recurse
+      # preorder goes from parent -> left -> right
+      list << { key: currNode.key, value: currNode.value }
+
+      preorder_helper(currNode.left, list)
+    
+      preorder_helper(currNode.right, list)
+
+      return list
+    end
+
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n) assuming a balanced tree, else O(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(currNode, list)
+
+    if currNode.nil?
+      # base case
+      return list
+
+    else
+      # recurse
+      # postorder goes from left -> right -> parent
+
+      postorder_helper(currNode.left, list)
+    
+      postorder_helper(currNode.right, list)
+
+      list << { key: currNode.key, value: currNode.value }
+
+      return list
+    end
+
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n) assuming a balanced tree, else O(n)
   def height
-    raise NotImplementedError
+    return subtree_height(root)
+  end
+
+  def subtree_height(currNode)
+    if currNode.nil?
+      return 0
+    else
+      if !currNode.left && !currNode.right
+        return 1
+      else
+        # set initial conditions for the subtrees  
+        leftHeight = 0
+        rightHeight = 0
+
+        if currNode.left 
+          leftHeight = 1 + subtree_height(currNode.left)
+        end
+        
+        if currNode.right
+          rightHeight = 1 + subtree_height(currNode.right)
+        end 
+
+        if leftHeight >= rightHeight 
+          return leftHeight
+        else
+          return rightHeight
+        end
+      end
+      
+    end
   end
 
   # Optional Method
   # Time Complexity: 
   # Space Complexity: 
   def bfs
-    raise NotImplementedError
+    return [] if !root
+
+    listOfNodes = [root]
+    answer = [{key: root.key, value: root.value} ]
+    currIndex = 0
+
+    while listOfNodes[currIndex]
+      puts "\nlist is currenty... "
+      printNodesInList(listOfNodes)
+
+      currNode = listOfNodes[currIndex]
+      puts "looking at listOfNodes: #{listOfNodes}"
+      if currNode.left
+        listOfNodes << currNode.left 
+        answer << {key: currNode.left.key, value: currNode.left.value} 
+      end
+      if currNode.right
+        listOfNodes << currNode.right
+        answer << {key: currNode.right.key, value: currNode.right.value} 
+      end
+      currIndex += 1
+    end
+    
+    return answer
+  end
+
+  def printNodesInList(list)
+    currIndex = 0
+    while list[currIndex]
+      puts "\t#{list[currIndex].key}/#{list[currIndex].value}"
+      currIndex += 1
+    end
   end
 
+  # Optional Method, IDK, i think mine is unnecessarily complicated
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n)???
+  def delete(key)
```

Comment:
This seems to work, but it may lead to an unbalanced tree because of how you're adding nodes back in.  I'm not certain from looking at this, but I suspect that to be the case.